components/image-controller/production/stone-prd-rh01/resources/image_pruner/prune_images.py

```diff
@@ -0,0 +1,182 @@
+from datetime import datetime
+from requests.exceptions import HTTPError
+
+logging.basicConfig(
+    format="%(asctime)s - %(levelname)s - %(message)s", level=logging.INFO
+)
+LOGGER = logging.getLogger(__name__)
+QUAY_API_URL = "https://quay.io/api/v1"
+DAY_OLD_TS = int(datetime.now().timestamp()) - (60 * 60 * 24)
+KEEP_MAX = 3
+
+ImageRepo = Dict[str, Any]
+
+
+def get_quay_tags(quay_token: str, namespace: str, name: str) -> ImageRepo:
+    next_page = None
+    resp: HTTPResponse
+
+    all_tags = []
+    while True:
+        query_args = {"limit": 100, "onlyActiveTags": True}
+        if next_page is not None:
+            query_args["page"] = next_page
+
+        api_url = f"{QUAY_API_URL}/repository/{namespace}/{name}/tag/?{urlencode(query_args)}"
+        request = Request(api_url, headers={
+            "Authorization": f"Bearer {quay_token}",
+        })
+
+        with urlopen(request) as resp:
+            if resp.status != 200:
+                raise RuntimeError(resp.reason)
+            json_data = json.loads(resp.read())
+
+        tags = json_data.get("tags", [])
+        all_tags.extend(tags)
+
+        if not tags:
+            LOGGER.debug("No tags found.")
+            break
+
+        page = json_data.get("page", None)
+        additional = json_data.get("has_additional", False)
+
+        if additional:
+            next_page = page + 1
+        else:
+            break
+
+    return all_tags
+
+
+def delete_image_tag(quay_token: str, namespace: str, name: str, tag: str) -> None:
+    api_url = f"{QUAY_API_URL}/repository/{namespace}/{name}/tag/{tag}"
+    request = Request(api_url, method="DELETE", headers={
+        "Authorization": f"Bearer {quay_token}",
+    })
+    resp: HTTPResponse
+
+    while True:
+        try:
+            with urlopen(request) as resp:
+                if resp.status != 200 and resp.status != 204 and resp.status != 404:
+                    raise RuntimeError(resp.reason)
+                else:
+                    break
+        except HTTPError as ex:
+            LOGGER.info("exception: %s", ex)
+
+
+def remove_leftover_tags(tags: List[Dict[str, Any]], quay_token: str, namespace: str, name: str,
+                         dry_run: bool = False) -> None:
+    tag_regex = re.compile(r"^sha256-([0-9a-f]+)(\.sbom|\.att|\.src|\.sig)$")
+
+    # remove att/sbom/src/sig for which is missing manifest digest
+    image_digests = [image["manifest_digest"] for image in tags]
+
+    for tag in tags:
+        if (match := tag_regex.match(tag["name"])) is not None:
+            if f"sha256:{match.group(1)}" not in image_digests:
+                if dry_run:
+                    LOGGER.info("Leftover image %s from %s/%s should be removed", tag["name"], namespace, name)
+                else:
+                    LOGGER.info("Removing leftover image %s from %s/%s", tag["name"], namespace, name)
+                    delete_image_tag(quay_token, namespace, name, tag["name"])
```

Comment:
>  * Is it possible to reuse the existing prune_images.py script under image-controller by making it more configurable?

> I don't want to mix those 2 scripts, they are doing different kind of pruning

_Are_ they doing different kinds of pruning? Isn't this exactly what the existing pruner does?

Reply:
yeah one is doing ONLY leftovers other is doing more, and like I've said bellow those scripts are for different purposes